### PR TITLE
Add support for Readthedocs

### PR DESCRIPTION
**Pull request by WillTrojak**
This adds support for documentation to viewed by Read The Docs. Changes include:

- Modifications to theme and style
- Restructuring of docs
- Added RTD pip requirements
- Added RTD yaml file

For this to be fully integrated webhooks will need to be setup by the admins. See: [https://docs.readthedocs.io/en/stable/webhooks.html](https://docs.readthedocs.io/en/stable/webhooks.html)

An online test of the documentation can be seen at: [https://pyfr-rtd-test.readthedocs.io/en/latest/index.html](https://pyfr-rtd-test.readthedocs.io/en/latest/index.html)

### CODE REVIEW

File: doc/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Sphinx >= 3.5.3
+graphviz
+mkdocs
+sphinx-rtd-theme
+sphinxcontrib-contentui
+sphinxcontrib-fancybox
+
+appdirs >= 1.4.0
```
Comment (FreddieWitherden) [inline doc/requirements.txt:8]: Are these bits all needed?  And how careful do we need to be about keeping them in sync with setup.py?

File: doc/src/conf.py [hunk 2]
```diff
@@ -31,7 +31,10 @@
     'sphinx.ext.graphviz',
     'sphinx.ext.inheritance_diagram',
     'sphinx.ext.mathjax',
```
Comment (FreddieWitherden) [inline doc/src/conf.py:33]: Is this needed?

File: doc/src/conf.py [hunk 3]
```diff
@@ -54,7 +58,11 @@
     penwidth='0.3')
 
 # The names of modules to mock and hence avoid import.
-autodoc_mock_imports = ['pyopencl']
+autodoc_mock_imports = ['pyopencl',
+                        'h5py',
+                        'mpi4py',
+                        'pycuda'
```
Comment (FreddieWitherden) [inline doc/src/conf.py:64]: pycuda?

File: doc/src/conf.py [hunk 4]
```diff
@@ -120,12 +128,32 @@
 
 # The theme to use for HTML and HTML Help pages.  See the documentation for
 # a list of builtin themes.
-html_theme = 'classic'
+# html_theme = 'classic'
```
Comment (FreddieWitherden) [inline doc/src/conf.py:131]: Remove?

File: doc/src/conf.py [hunk 5]
```diff
@@ -120,12 +128,32 @@
...
+html_favicon = "../fig/pyfr_favicon.png"
+
+
 
 # Theme options are theme-specific and customize the look and feel of a theme
 # further.  For a list of options available for each theme, see the
 # documentation.
-#html_theme_options = {}
+html_theme_options = {
+    'analytics_id': 'G-RPKQBYNX01',  #  Provided by Google in your dashboard
```
Comment (FreddieWitherden) [inline doc/src/conf.py:142]: Do we need this?

File: doc/src/examples.rst [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+6. Run pyfr to solve the Euler equations on the mesh, generating a
+   series of PyFR solution files called ``euler_vortex_2d*.pyfrs``::
+
+        mpiexec -n 2 pyfr run -b cuda -p euler_vortex_2d.pyfrm euler_vortex_2d.ini
+
+7. Run pyfr on the solution file ``euler_vortex_2d-100.0.pyfrs``
+   converting it into an unstructured VTK file called
+   ``euler_vortex_2d-100.0.vtu``. Note that in order to visualise the
+   high-order data, each high-order element is sub-divided into smaller
+   linear elements. The level of sub-division is controlled by the
```
Comment (FreddieWitherden) [inline doc/src/examples.rst:44]: Remove the subdivision stuff?

File: doc/src/installation.rst [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+************
+
+Quick-start
+===========
+
+PyFR |release| can be installed using `pip <https://pypi.python.org/pypi/pip>`_
+and `virtualenv <https://pypi.python.org/pypi/virtualenv>`_, as shown in the
+quick-start guides below.
+
+Alternatively, PyFR |release| can be installed from `source <http://www.pyfr.org/download.php>`_, see :ref:`compile-from-source`.
```
Comment (FreddieWitherden) [inline doc/src/installation.rst:14]: Wrap line.

File: doc/src/installation.rst [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    cc = gcc
+
+This has been tested on Ubuntu 18.04.
+
+.. _compile-from-source:
+
+Compiling from source
+=====================
+
+PyFR can be obtained `here <http://www.pyfr.org/download.php>`_. To install the software from source, use the provided ``setup.py`` installer or add the root PyFR directory to ``PYTHONPATH`` using::
```
Comment (FreddieWitherden) [inline doc/src/installation.rst:88]: Wrap line.

File: doc/src/installation.rst [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+PyFR |release| has a hard dependency on Python 3.6+ and the following
+Python packages:
+
+1. `appdirs <https://github.com/ActiveState/appdirs>`_ >= 1.4.0
+2. `gimmik <https://github.com/vincentlab/GiMMiK>`_ >= 2.0
+3. `h5py <http://www.h5py.org/>`_ >= 2.6
+4. `mako <http://www.makotemplates.org/>`_ >= 1.0.0
+5. `mpi4py <http://mpi4py.scipy.org/>`_ >= 2.0
+6. `numpy <http://www.numpy.org/>`_ >= 1.8
```
Comment (FreddieWitherden) [inline doc/src/installation.rst:108]: I believe we need a newer version; check setup.py.

File: doc/src/installation.rst [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+The backend requires:
+
+1. `ROCm <https://rocmdocs.amd.com/en/latest/>`_ >= 4.0
+2. `rocBLAS <https://github.com/ROCmSoftwarePlatform/rocBLAS>`_ >= 2.32.0
+
+OpenCL Backend
+^^^^^^^^^^^^^^
+
+The OpenCL backend targets a range of accelerators including GPUs from
+AMD and NVIDIA. The backend requires:
```
Comment (FreddieWitherden) [inline doc/src/installation.rst:135]: Include Intel here, too?

File: doc/src/conf.py (hunk 2 above)
Comment (WillTrojak) [reply doc/src/conf.py:33]: I do use Mathjax in one place, I don't think there's any harm in keeping it.

File: doc/src/conf.py (hunk 3 above)
Comment (WillTrojak) [reply doc/src/conf.py:64]: This is leftover from trying to get the backwards compatibility to work, will remove

File: doc/src/conf.py (hunk 5 above)
Comment (WillTrojak) [reply doc/src/conf.py:142]: No

File: doc/src/user_guide.rst [hunk 11]
```diff
@@ -448,13 +298,20 @@ Sets constants used in the simulation
 
    *float*
 
+Other constant may be set by the user which can then be used throughout the ``.ini`` file.
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:301]: Line length.

File: doc/src/user_guide.rst [hunk 12]
```diff
@@ -712,1034 +569,975 @@ Example::
...
 
-[soln-plugin-writer]
-^^^^^^^^^^^^^^^^^^^^
-Periodically write the solution to disk in the pyfrs format.
-Parameterised with
+Periodic Boundary Conditions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-1. ``dt-out`` --- write to disk every ``dt-out`` time units:
+Simple periodic boundary conditions are supported; however, there behaviour
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:903]: their

File: doc/src/user_guide.rst [hunk 13]
```diff
@@ -712,1034 +569,975 @@ Example::
...
-7. Visualise the unstructured VTK file in `Paraview
-   <http://www.paraview.org/>`_
+4. ``mode`` --- output file accumulation mode:
 
-.. figure:: ../fig/couette_flow_2d/couette_flow_2d.png
-   :width: 450px
-   :figwidth: 450px
-   :alt: couette flow
-   :align: center
+    ``continuous`` | ``windowed``
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1446]: Do we want to say something more here?

**Comment by GonzaloSaez** [comment]
Would it be possible to add to the `doc` how to use several plugins of the same type within a unique simulation? i.e.

```
To use several plugins of the same type, one may add a labelling to each plugin as follows

[soln-plugin-writer]
...

[soln-plugin-writer-2]
...
```

File: doc/src/user_guide.rst [hunk 14]
```diff
@@ -1220,6 +1220,14 @@ Plugins
 -------
 
 Plugins allow for powerful additional functionality to be swapped in and out.
+It is possible to load multiple instances of the same plugin by appending an
+integer tag, for example::
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1224]: So the tag can be more than an integer.  (Also see the fluidforce plugin which uses the tag explicitly.)

File: doc/src/examples.rst [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        pyfr run -b cuda -p couette_flow_2d.pyfrm couette_flow_2d.ini
+
+6. Run pyfr on the solution file ``couette_flow_2d-040.pyfrs``
+   converting it into an unstructured VTK file called
+   ``couette_flow_2d-040.vtu``. Note that in order to visualise the
+   high-order data, each high-order element is sub-divided into smaller
+   linear elements. The level of sub-division is controlled by the
+   integer at the end of the command::
+
+        pyfr export couette_flow_2d.pyfrm couette_flow_2d-040.pyfrs couette_flow_2d-040.vtu -d 4
```
Comment (FreddieWitherden) [inline doc/src/examples.rst:96]: Remove subdivision.

File: doc/src/examples.rst [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+        mpiexec -n 2 pyfr run -b cuda -p euler_vortex_2d.pyfrm euler_vortex_2d.ini
+
+7. Run pyfr on the solution file ``euler_vortex_2d-100.0.pyfrs``
+   converting it into an unstructured VTK file called
+   ``euler_vortex_2d-100.0.vtu``. To visualise the results using 
+   subdivision, a control integer is added at the end of the 
+   command::
+
+        pyfr export euler_vortex_2d.pyfrm euler_vortex_2d-100.0.pyfrs euler_vortex_2d-100.0.vtu -d 4
```
Comment (FreddieWitherden) [inline doc/src/examples.rst:46]: Remove subdivision.

File: doc/src/examples.rst [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        pyfr run -b cuda -p inc_cylinder_2d.pyfrm inc_cylinder_2d.ini
+
+6. Run pyfr on the solution file ``inc_cylinder_2d-75.00.pyfrs``
+   converting it into an unstructured VTK file called
+   ``inc_cylinder_2d-75.00.vtu``. Note that in order to visualise the
+   high-order data, each high-order element is sub-divided into smaller
+   linear elements. The level of sub-division is controlled by the
+   integer at the end of the command::
+
+        pyfr export inc_cylinder_2d.pyfrm inc_cylinder_2d-75.00.pyfrs inc_cylinder_2d-75.00.vtu -d 4
```
Comment (FreddieWitherden) [inline doc/src/examples.rst:146]: Remove subdivision.

**Comment by FreddieWitherden** [comment]
Also, I do not think that the integration plugin is documented?

File: doc/src/user_guide.rst [hunk 18]
```diff
@@ -4,181 +4,16 @@
 User Guide
 **********
 
-Getting Started
-===============
+This will provide infomation on :ref:`running-pyfr` and how calcualtion
```
Comment (pv101) [inline doc/src/user_guide.rst:7]: Remove 'calcualtion' ?

File: doc/requirements.txt (hunk 1 above)
Comment (pv101) [reply doc/requirements.txt:8]: Was this resolved?

File: doc/src/conf.py [hunk 19]
```diff
@@ -257,7 +285,7 @@
 #  dir menu entry, description, category)
 texinfo_documents = [
   ('index', 'PyFR', u'PyFR Documentation',
-   u'Imperial College London', 'PyFR', 'One line description of project.',
+   u'Imperial College London', 'PyFR', 'A high-order flux reconstruction solver.',
```
Comment (pv101) [inline doc/src/conf.py:288]: Flux Reconstruction

File: doc/src/examples.rst (hunk 15 above)
Comment (pv101) [reply doc/src/examples.rst:96]: Is there an argument for leaving the subdivision in so that the .vtu files will work with older version of Paraview that don't support high-order cells?

File: doc/src/examples.rst (hunk 17 above)
Comment (pv101) [reply doc/src/examples.rst:146]: Is there an argument for leaving the subdivision in so that the .vtu files will work with older version of Paraview that don't support high-order cells?

File: doc/src/examples.rst (hunk 15 above)
Comment (FreddieWitherden) [reply doc/src/examples.rst:96]: On balance probably not.  That one argument gives rise to two additional explanatory sentences; one of which is no longer true (we can visualise high-order data directly and, indeed, it is the recommended solution by default).

Comment (pv101) [reply doc/src/examples.rst:96]: OK, I agree with removing the -d 4 (along with the explanatory sentences).

File: doc/src/installation.rst [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    virtualenv --python=python3 ENV3
+    source ENV3/bin/activate
+
+Finally install PyFR with `pip <https://pypi.python.org/pypi/pip>`_ in the
+virtual environment::
+
+    pip install pyfr
+
+This concludes the installation. In order to run PyFR with the OpenMP backend
+(see :ref:`running-pyfr`), use the following settings in the configuration file
```
Comment (pv101) [inline doc/src/installation.rst:42]: You could make the reference to the config file actually link to the relevant page with :ref:`...` ?

File: doc/src/installation.rst [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+(``.ini``)::
+
+    [backend-openmp]
+    cc = gcc-8
+    
+Note the version of the compiler which must support the ``openmp`` flag.
```
Comment (pv101) [inline doc/src/installation.rst:48]: Is this a Mac specific thing?

File: doc/src/installation.rst [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+    python3 -m virtualenv ENV3
+    source ENV3/bin/activate
+
+Finally install PyFR with `pip <https://pypi.python.org/pypi/pip>`_ in the
+virtual environment::
+
+    pip install pyfr
+
+This concludes the installation. In order to run PyFR with the OpenMP backend
+(see :ref:`running-pyfr`), use the following settings in the configuration file
```
Comment (pv101) [inline doc/src/installation.rst:76]: You can also make this a link to the config file docs.

File: doc/src/installation.rst [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+
+Quick-start
+===========
+
+PyFR |release| can be installed using `pip <https://pypi.python.org/pypi/pip>`_
+and `virtualenv <https://pypi.python.org/pypi/virtualenv>`_, as shown in the
+quick-start guides below.
+
+Alternatively, PyFR |release| can be installed from 
+`source <http://www.pyfr.org/download.php>`_, see :ref:`compile-from-source`.
```
Comment (FreddieWitherden) [inline doc/src/installation.rst:15]: We probably want to update this URL.

File: doc/src/user_guide.rst (hunk 18 above)
Comment (pv101) [reply doc/src/user_guide.rst:7]: In fact, I would remove this whole bit of text: "This will provide infomation on Running PyFR and how calcualtion the Configuration File (.ini) is setup. For information on how to install PyFR see Installation."

File: doc/src/installation.rst (hunk 21 above)
Comment (FreddieWitherden) [reply doc/src/installation.rst:48]: Yes, I think by default macOS does not come with a compiler that supports OpenMP, although someone might want to check the status of 10.15.

File: doc/src/installation.rst (hunk 23 above)
Comment (pv101) [reply doc/src/installation.rst:15]: Good catch. We should point to the Github repo now.

File: doc/src/installation.rst [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+    cc = gcc
+
+This has been tested on Ubuntu 18.04.
+
+.. _compile-from-source:
+
+Compiling from source
+=====================
+
+PyFR can be obtained `here <http://www.pyfr.org/download.php>`_. To install the
```
Comment (pv101) [inline doc/src/installation.rst:89]: This should point to the Github repo as well.

File: doc/src/user_guide.rst [hunk 25]
```diff
@@ -238,23 +73,31 @@ The following commands are available from the ``pyfr`` program:
    found in the solution file.
 
 Running in Parallel
-^^^^^^^^^^^^^^^^^^^
+-------------------
 
 ``pyfr`` can be run in parallel. To do so prefix ``pyfr`` with
```
Comment (pv101) [inline doc/src/user_guide.rst:78]: I think this first ``pyfr`` can just be PyFR

File: doc/src/user_guide.rst [hunk 26]
```diff
@@ -238,23 +73,31 @@ The following commands are available from the ``pyfr`` program:
...
 in the `INI <http://en.wikipedia.org/wiki/INI_file>`_ format.
 Parameters are grouped into sections. The roles of each section and
-their associated parameters are described below.
+their associated parameters are described below. Note that both ``;`` and 
+``#`` may be used as comment characters.
+
+Backends
+--------
+
+There are several configuration setting for the backends. The base
```
Comment (pv101) [inline doc/src/user_guide.rst:97]: Rework English, is hard to follow, also correct spelling: "The base 
section ``[backend]`` is required, whereas PyFR has internal defualts for the 
other hardware specific backend settings; however, for best perfomance it is 
advisable to configure those applicable for the target hardware."

File: doc/src/user_guide.rst [hunk 27]
```diff
@@ -411,6 +254,13 @@ Example::
     cblas= example/path/libBLAS.dylib
     cblas-type = parallel
 
+Systems
+-------
+
+These sections of the input file broadly setup and control the physical
```
Comment (pv101) [inline doc/src/user_guide.rst:260]: Drop 'broadly'

File: doc/src/user_guide.rst [hunk 28]
```diff
@@ -712,1034 +570,997 @@ Example::
...
+    [soln-bcs-bcwallupper]
+    type = no-slp-isot-wall
+    cpTw = 10.0
+    u = 1.0
 
-[soln-plugin-writer]
-^^^^^^^^^^^^^^^^^^^^
-Periodically write the solution to disk in the pyfrs format.
-Parameterised with
+Periodic Boundary Conditions
```
Comment (pv101) [inline doc/src/user_guide.rst:901]: Drop this heading (but keep the subsequent text).

File: doc/src/user_guide.rst [hunk 29]
```diff
@@ -712,1034 +570,997 @@ Example::
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-1. ``dt-out`` --- write to disk every ``dt-out`` time units:
+Simple periodic boundary conditions are supported; however, their behaviour
+is not controlled through the ``.ini`` file, instead it is handled at
+the mesh generation stage. Two faces may be taged with
+``periodic_l_x`` and ``periodic_r_x``, where ``x`` is a unique
+integer for the pair of boundaries. Currently, only periodicity in a
+single cardinal direction is supported, for example, the planes
+:math:`(x,y,0)` and :math:`(x,y,10)`.
```
Comment (pv101) [inline doc/src/user_guide.rst:910]: Is there a specific reason why these are :math: formatted?

File: doc/src/user_guide.rst [hunk 30]
```diff
@@ -411,6 +254,13 @@ Example::
     cblas= example/path/libBLAS.dylib
```
Comment (semi-h) [inline doc/src/user_guide.rst:254]: We don't support cblas any more with the openmp kernel

File: doc/src/user_guide.rst [hunk 31]
```diff
@@ -712,1034 +570,997 @@ Example::
...
+    v = z
+    w = 1.0
+    p = 1.0/(1.0+x)
 
-1. ``nsteps`` --- check every ``nsteps``:
+Nodal Point Sets
+----------------
 
-    *int*
+Flux reconstruction is a high-order nodal spectral element method, and
```
Comment (pv101) [inline doc/src/user_guide.rst:954]: I would replace with something shorter, like: "Solution point sets must be specified for each element type that is used and flux point sets must be specified for each interface type that is used. If anti-aliasing is enabled then quadrature point for each element and interface type that is used must also be specified."

File: doc/src/user_guide.rst [hunk 32]
```diff
@@ -712,1034 +570,997 @@ Example::
...
+4. ``mode`` --- output file accumulation mode:
 
-        pyfr export couette_flow_2d.pyfrm couette_flow_2d-040.pyfrs couette_flow_2d-040.vtu -d 4
+    ``continuous`` | ``windowed``
 
-7. Visualise the unstructured VTK file in `Paraview
-   <http://www.paraview.org/>`_
+    The default is ``windowed``. Continuous uses an accumulation over the entire
+    averaging period currently available, whereas windowed averages only over 
+    the number of steps. Oweing to the associativity and commutativity
```
Comment (pv101) [inline doc/src/user_guide.rst:1459]: What do we mean by 'the number of steps'?

Comment (pv101) [reply doc/src/user_guide.rst:1459]: Also what does 'currently available' mean?

File: doc/src/examples.rst (hunk 15 above)
Comment (GiorgioGiangaspero) [reply doc/src/examples.rst:96]: Probably worth mentioning here to increase the 'Nonlinear Subdivision level' in Paraview otherwise I think the rendering will not show the high-order solution by default. We also may need to update the pictures for each example.

File: doc/src/user_guide.rst [hunk 33]
```diff
@@ -712,1034 +560,995 @@ Example::
...
 
-1. ``nsteps`` --- check every ``nsteps``:
+Nodal Point Sets
+----------------
 
-    *int*
+Solution point sets must be specified for each element type that is used and
+flux point sets must be specified for each interface type that is used. If
+anti-aliasing is enabled then quadrature point for each element and interface
+type that is used must also be specified. For example, a 3D mesh comprised only
```
Comment (pv101) [inline doc/src/user_guide.rst:944]: For example, a 3D mesh comprised only of prisms requires a solution point set for prism elements and flux point sets for quadrilateral and triangular interfaces.

File: doc/src/user_guide.rst [hunk 34]
```diff
@@ -712,1034 +560,995 @@ Example::
...
+    p = 1.0/(1.0+x)
 
-1. ``nsteps`` --- check every ``nsteps``:
+Nodal Point Sets
+----------------
 
-    *int*
+Solution point sets must be specified for each element type that is used and
+flux point sets must be specified for each interface type that is used. If
+anti-aliasing is enabled then quadrature point for each element and interface
```
Comment (pv101) [inline doc/src/user_guide.rst:943]: quadrature point sets

File: doc/src/user_guide.rst [hunk 35]
```diff
@@ -712,1034 +560,995 @@ Example::
...
-   <http://www.paraview.org/>`_
+Additional Information
+----------------------
 
-.. figure:: ../fig/inc_cylinder_2d/inc_cylinder_2d.png
-   :width: 450px
-   :figwidth: 450px
-   :alt: couette flow
-   :align: center
+The ini file format is very versatile. A feature that can be useful in
```
Comment (pv101) [inline doc/src/user_guide.rst:1547]: ini -> INI (and link to the INI format page) maybe given usage here?

File: doc/src/user_guide.rst [hunk 36]
```diff
@@ -712,1034 +560,995 @@ Example::
...
+defining initial conditions is the substitution feature and this is
+demonstrated in the :ref:`integrate-plugin` example.
 
-   Colour map of velocity magnitude distribution at 75 time units.
+To prevent situations where you have solutions files for unknown
+configurations, the contents of the ini file is added as an attribute
```
Comment (pv101) [inline doc/src/user_guide.rst:1552]: Should this ini be .ini ? We seem to use a mix of ini .ini ``.ini`` and INI. Maybe quick check for consistency?

File: doc/requirements.txt (hunk 1 above)
Comment (FreddieWitherden) [reply doc/requirements.txt:8]: Bump.
